File: contests/itmo/masters-3-sem/002/A_Brackets.py
import sys


def is_opening_bracket(c):
    return c in '([{'


def closing_bracket(c):
    if c == '(':
        return ')'
    elif c == '[':
        return ']'
    else:
        return '}'


def check_brackets(s):
    st = []

    for val in s:
        if is_opening_bracket(val):
            st.append(val)
        else:
            if len(st) == 0 or val != closing_bracket(st[-1]):
                return False
            st.pop()

    return len(st) == 0


if __name__ == "__main__":
    s = sys.stdin.read().strip()

    if check_brackets(s):
        print("YES")
    else:
        print("NO")

File: contests/itmo/masters-3-sem/002/A_Brackets.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func isOpeningBracket(r rune) bool {
	return r == '(' || r == '[' || r == '{'
}

func closingBracket(r rune) rune {
	if r == '(' {
		return ')'
	} else if r == '[' {
		return ']'
	} else {
		return '}'
	}
}

func checkBrackets(s string) bool {
	stack := make([]rune, 0)

	for _, val := range s {
		if isOpeningBracket(val) {
			stack = append(stack, val)
		} else {
			if len(stack) == 0 || val != closingBracket(stack[len(stack)-1]) {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()

	s := scanner.Text()

	if checkBrackets(s) {
		fmt.Println("YES")
	} else {
		fmt.Println("NO")
	}
}
